core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/kerberos/KeytabConfigurationHandler.java

```diff
@@ -0,0 +1,119 @@
+        return EventSelectorUtil.selector(KeytabConfigurationRequest.class);
+    }
+
+    @Override
+    public void accept(Event<KeytabConfigurationRequest> keytabConfigurationRequestEvent) {
+        Long stackId = keytabConfigurationRequestEvent.getData().getResourceId();
+        Selectable response;
+        try {
+            Stack stack = stackService.getByIdWithListsInTransaction(stackId);
+            Optional<KerberosConfig> kerberosConfigOptional = kerberosConfigService.get(stack.getEnvironmentCrn());
+            // TODO remove YARN check when FreeIPA registration is ready
+            if (!CloudPlatform.YARN.name().equals(stack.cloudPlatform())
+                    && kerberosConfigOptional.isPresent() && kerberosDetailService.isIpaJoinable(kerberosConfigOptional.get())) {
+                GatewayConfig primaryGatewayConfig = gatewayConfigService.getPrimaryGatewayConfig(stack);
+                ServiceKeytabResponse serviceKeytabResponse = getServiceKeytabResponse(stack, primaryGatewayConfig);
+                KeytabModel keytabModel = buildKeytabModel(serviceKeytabResponse);
+                hostOrchestrator.uploadKeytabs(List.of(primaryGatewayConfig), Set.of(keytabModel),
+                        ClusterDeletionBasedExitCriteriaModel.clusterDeletionBasedModel(stackId, stack.getCluster().getId()));
+            }
+            response = new KeytabConfigurationSuccess(stackId);
+        } catch (Exception e) {
+            LOGGER.info("Error during keytab configuration, stackId: " + stackId, e);
+            response = new KeytabConfigurationFailed(stackId, e);
+        }
+        eventBus.notify(response.selector(), new Event<>(keytabConfigurationRequestEvent.getHeaders(), response));
+    }
+
+    private KeytabModel buildKeytabModel(ServiceKeytabResponse serviceKeytabResponse) {
+        String keytabInBase64 = secretService.getByResponse(serviceKeytabResponse.getKeytab());
+        byte[] keytab = Base64.getDecoder().decode(keytabInBase64.getBytes(StandardCharsets.UTF_8));
+        String principal = secretService.getByResponse(serviceKeytabResponse.getServicePrincial());
+        return new KeytabModel("CM", "/etc/cloudera-scm-server", "cmf.keytab", principal, keytab);
+    }
+
+    private ServiceKeytabResponse getServiceKeytabResponse(Stack stack, GatewayConfig primaryGatewayConfig)
+            throws Exception {
+        ServiceKeytabRequest request = new ServiceKeytabRequest();
+        request.setEnvironmentCrn(stack.getEnvironmentCrn());
+        request.setServerHostName(primaryGatewayConfig.getHostname());
+        request.setServiceName("CM");
+        request.setDoNotRecreateKeytab(Boolean.TRUE);
+        RoleRequest roleRequest = new RoleRequest();
+        roleRequest.setRoleName("hadoopadminrole-" + stack.getName());
+        roleRequest.setPrivileges(Set.of("Service Administrators"));
+        request.setRoleRequest(roleRequest);
+        return kerberosMgmtV1Endpoint.generateServiceKeytab(request);
```

Comment:
Can you use api "getServiceKeytab", instead? When you use this getServiceKeytab, there is no need to pass doNotRecreateKeytab in the request.

Reply:
I was going to comment elsewhere but I will add it here. There is a get and post method for /v1/kerberosmgmt/servicekeytab. The intent was that post method will always create a new keytab and if there are existing keytabs it will always invalidate them. The get method will never invalidate the existing keytabs but it will also never create a keytab.

It seems that doNotRecreateKeytab seems to enable the post method to be used to attempt to encapsulate logic that a client already can do try a "get" and if it fails then do a "post". The problem I see with hiding this logic in the FreeIPA management service is that is that it may actually lead to invalidations in some race cases. Consider what happens if this is called twice in a very short period of time. When we move to an HA environment, the time period of race condition expose is enlarged. In HA, FreeIPA operates with multimaster. The later timestamp will win. This means the second request can actually cause the regeneration of the keytab and thus cause the invalidation of the first.

So I think doNotRecreateKeytab will mislead the caller of the API and lead to situations that are not possible to guarantee (since its not supported by the underlying FreeIPA APIs).

Reply:
the main reason for this to avoid 2 calls from CB to FreeIPA. If we don't have this option, then we have to make a call with get to see if there is already a keytab for the service. If no, we can call the create. This way our call from CB is idempotent, we can call it many times.

From CB side I can't see the issue of calling this method in a short time causing race condition, but if it's an issue from other services I can rewrite it to do a call with get.

Reply:
If we try to get first then post we could encounter the same race condition, aren't we?

Reply:
@keyki , you are correct. The race condition exists no matter what we do. I am just suggesting that we not try to pretend there isn't a race condition. Regardless of `doNotRecreateKeytab`, the race condition is present. This API will be exposed to our internal workloads and I just don't want them to be surprised when `doNotRecreateKeytab=True` actually recreates a keytab and thus invalidates prior keytabs. I guess I am OK with adding the `doNotRecreateKeytab` flag as long as the API documentation warns that it is a "best effort" which is not guaranteed and may actually invalidate prior keytabs.

As to the short time, what happens if there is an HA scenario and a temporary network partition between the FreeIPA masters? Two post requests could be mapped to different FreeIPA servers. They are multi-master so there will be two valid responses. But after the network partition is healed, only one of the two keytabs will be usable. This network partition could potentially last longer than a "short period of time".

But I understand the desire to minimize the number of round trips requests, so can we just reword the documentation a little bit?

Reply:
ok, I will modify the docs according to your suggestion